rskj-core/src/main/java/co/rsk/peg/pegininstructions/PeginInstructionsProvider.java

```diff
@@ -0,0 +1,85 @@
+package co.rsk.peg.pegininstructions;
+
+import co.rsk.bitcoinj.core.BtcTransaction;
+import co.rsk.bitcoinj.script.ScriptChunk;
+import co.rsk.bitcoinj.script.ScriptOpCodes;
+import co.rsk.peg.NoOpReturnException;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PeginInstructionsProvider {
+
+    private static final Logger logger = LoggerFactory.getLogger(PeginInstructionsProvider.class);
+
+    public Optional<PeginInstructions> buildPeginInstructions(BtcTransaction btcTx) throws
+        PeginInstructionsException {
+
+        PeginInstructions peginInstructions;
+        byte[] opReturnOutputData;
+
+        try {
+            opReturnOutputData = extractOpReturnData(btcTx);
+        } catch (NoOpReturnException e) {
+            return Optional.empty();
+        }
+
+        if (opReturnOutputData == null || opReturnOutputData.length < 22) {
+            String message = "Invalid OP_RETURN data found";
+            logger.debug("[buildPeginInstructions] {}", message);
+            throw new PeginInstructionsException(message);
+        }
+
+        int protocolVersion = PeginInstructionsBase.extractProtocolVersion(opReturnOutputData);
+
+        switch (protocolVersion) {
+            case 1:
+                PeginInstructionsVersion1 peginInstructionsVersion1 = new PeginInstructionsVersion1(btcTx.getParams());
+                peginInstructionsVersion1.parse(opReturnOutputData);
+                peginInstructions = peginInstructionsVersion1;
+                break;
+            default:
+                logger.debug("[buildPeginInstructions] Invalid protocol version given");
+                throw new PeginInstructionsException("Invalid protocol version");
+        }
+
+        return Optional.of(peginInstructions);
+    }
+
+    protected static byte[] extractOpReturnData(BtcTransaction btcTx)
+        throws PeginInstructionsException {
+        byte[] data = new byte[]{};
+        int opReturnOccurrences = 0;
+
+        logger.info("[extractOpReturnData] Getting OP_RETURN data for btc tx: {}", btcTx.getHash());
+
+        for (int i=0; i<btcTx.getOutputs().size(); i++) {
+            List<ScriptChunk> chunksByOutput = btcTx.getOutput(i).getScriptPubKey().getChunks();
+            if (chunksByOutput.get(0).opcode == ScriptOpCodes.OP_RETURN) {
```

Comment:
I wonder if this could trigger an NPE

Reply:
No, because it is not possible for a BTC tx not tu have any outputs. At least one output will exist, if it is not OP_RETURN doesn't matter.